### PR TITLE
[runtime] If the initial pinvoke lookup fails, emit some code inside the managed-to-native wrapper to retry to lookup on every call until it succeeds. 

### LINKED ISSUE #35219: CoreCLR tests on Mono runtime : Failures in Interop/NativeLibrary* tests
Several failures seen when running the Interop/NativeLibrary* tests. These are CoreCLR tests  run on the mono runtime from the dotnet\runtime repo.
Repro : Command to run individual tests : From the dotnet\runtime repo src\mono\netcore run this command or relevant one for a particular test - make run-tests-coreclr CoreClrTest="bash ../../artifacts/tests/coreclr/OSX.x64.Debug/Regressions/coreclr/Interop/NativeLibraryResolveEvent/ResolveEventTests/ResolveEventTests.sh"

1. Interop/NativeLibraryResolveCallback/CallbackTests/CallbackTests.sh
2. Interop/NativeLibraryResolveCallback/CallbackStressTest/CallbackStressTest.sh
3. Interop/NativeLibraryResolveEvent/ResolveEventTests/ResolveEventTests.sh

Full stacks below : 

Interop/NativeLibraryResolveCallback/CallbackTests/CallbackTests.sh [FAIL]
          
          Return code:      1
          Raw output file:      /Users/samarthpatel/Documents/DotNetRunTime/runtime/artifacts/tests/coreclr/OSX.x64.Debug/Interop/NativeLibraryResolveCallback/Reports/Interop.NativeLibraryResolveCallback/CallbackTests/CallbackTests.output.txt
          Raw output:
          BEGIN EXECUTION
          /Users/samarthpatel/Documents/DotNetRunTime/runtime/artifacts/tests/coreclr/OSX.x64.Debug/Tests/Core_Root/corerun CallbackTests.dll ''
          Expected: 100
          Actual: 106
          END EXECUTION - FAILED
          Test Harness Exitcode is : 1
          To run the test:
          > set CORE_ROOT=/Users/samarthpatel/Documents/DotNetRunTime/runtime/artifacts/tests/coreclr/OSX.x64.Debug/Tests/Core_Root
          > /Users/samarthpatel/Documents/DotNetRunTime/runtime/artifacts/tests/coreclr/OSX.x64.Debug/Interop/NativeLibraryResolveCallback/CallbackTests/CallbackTests.sh
          Expected: True
          Actual:   False
          Stack Trace:
            /Users/samarthpatel/Documents/DotNetRunTime/runtime/artifacts/tests/coreclr/OSX.x64.Debug/TestWrappers/Interop.NativeLibraryResolveCallback/Interop.NativeLibraryResolveCallback.XUnitWrapper.cs(237,0): at Interop_NativeLibraryResolveCallback._CallbackTests_CallbackTests_._CallbackTests_CallbackTests_sh()
            /Users/samarthpatel/Documents/DotNetRunTime/runtime/src/mono/netcore/System.Private.CoreLib/src/System/Reflection/RuntimeMethodInfo.cs(339,0): at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
         

***************************************        Interop/NativeLibraryResolveCallback/CallbackStressTest/CallbackStressTest.sh [FAIL]
          
          Unhandled Exception:
          System.DllNotFoundException: NativeLib assembly:<unknown assembly> type:<unknown type> member:(null)
             at CallbackStressTest.Main() in /Users/samarthpatel/Documents/DotNetRunTime/runtime/src/coreclr/tests/src/Interop/NativeLibraryResolveCallback/CallbackStressTest.cs:line 152
          [ERROR] FATAL UNHANDLED EXCEPTION: System.DllNotFoundException: NativeLib assembly:<unknown assembly> type:<unknown type> member:(null)
             at CallbackStressTest.Main() in /Users/samarthpatel/Documents/DotNetRunTime/runtime/src/coreclr/tests/src/Interop/NativeLibraryResolveCallback/CallbackStressTest.cs:line 152
          
          Return code:      1
          Raw output file:      /Users/samarthpatel/Documents/DotNetRunTime/runtime/artifacts/tests/coreclr/OSX.x64.Debug/Interop/NativeLibraryResolveCallback/Reports/Interop.NativeLibraryResolveCallback/CallbackStressTest/CallbackStressTest.output.txt
          Raw output:
          BEGIN EXECUTION
          /Users/samarthpatel/Documents/DotNetRunTime/runtime/artifacts/tests/coreclr/OSX.x64.Debug/Tests/Core_Root/corerun CallbackStressTest.dll ''
          Expected: 100
          Actual: 1
          END EXECUTION - FAILED
          Test Harness Exitcode is : 1
          To run the test:
          > set CORE_ROOT=/Users/samarthpatel/Documents/DotNetRunTime/runtime/artifacts/tests/coreclr/OSX.x64.Debug/Tests/Core_Root
          > /Users/samarthpatel/Documents/DotNetRunTime/runtime/artifacts/tests/coreclr/OSX.x64.Debug/Interop/NativeLibraryResolveCallback/CallbackStressTest/CallbackStressTest.sh
          Expected: True
          Actual:   False
          Stack Trace:
            /Users/samarthpatel/Documents/DotNetRunTime/runtime/artifacts/tests/coreclr/OSX.x64.Debug/TestWrappers/Interop.NativeLibraryResolveCallback/Interop.NativeLibraryResolveCallback.XUnitWrapper.cs(140,0): at Interop_NativeLibraryResolveCallback._CallbackStressTest_CallbackStressTest_._CallbackStressTest_CallbackStressTest_sh()
            /Users/samarthpatel/Documents/DotNetRunTime/runtime/src/mono/netcore/System.Private.CoreLib/src/System/Reflection/RuntimeMethodInfo.cs(339,0): at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
          

***************************************        Interop/NativeLibraryResolveEvent/ResolveEventTests/ResolveEventTests.sh [FAIL]
          
          Return code:      1
          Raw output file:      /Users/samarthpatel/Documents/DotNetRunTime/runtime/artifacts/tests/coreclr/OSX.x64.Debug/Interop/NativeLibraryResolveEvent/Reports/Interop.NativeLibraryResolveEvent/ResolveEventTests/ResolveEventTests.output.txt
          Raw output:
          BEGIN EXECUTION
          /Users/samarthpatel/Documents/DotNetRunTime/runtime/artifacts/tests/coreclr/OSX.x64.Debug/Tests/Core_Root/corerun ResolveEventTests.dll ''
          Expected: 100
          Actual: 105
          END EXECUTION - FAILED
          Test Harness Exitcode is : 1
          To run the test:
          > set CORE_ROOT=/Users/samarthpatel/Documents/DotNetRunTime/runtime/artifacts/tests/coreclr/OSX.x64.Debug/Tests/Core_Root
          > /Users/samarthpatel/Documents/DotNetRunTime/runtime/artifacts/tests/coreclr/OSX.x64.Debug/Interop/NativeLibraryResolveEvent/ResolveEventTests/ResolveEventTests.sh
          Expected: True
          Actual:   False
          Stack Trace:
            /Users/samarthpatel/Documents/DotNetRunTime/runtime/artifacts/tests/coreclr/OSX.x64.Debug/TestWrappers/Interop.NativeLibraryResolveEvent/Interop.NativeLibraryResolveEvent.XUnitWrapper.cs(140,0): at Interop_NativeLibraryResolveEvent._ResolveEventTests_ResolveEventTests_._ResolveEventTests_ResolveEventTests_sh()
            /Users/samarthpatel/Documents/DotNetRunTime/runtime/src/mono/netcore/System.Private.CoreLib/src/System/Reflection/RuntimeMethodInfo.cs(339,0): at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)

### PR DESCRIPTION
**Pull request by vargaz**
Fixes #35219

### CODE REVIEW

**Comment by vargaz** [comment]
This needs some interpreter changes.

**Comment by srxqds** [comment]
is there an illegal effect for aot?

**Comment by vargaz** [comment]
AOT support is not implemented yet, it will fail the same way as before this PR.

**Comment by CoffeeFlux** [comment]
Worth noting this is unlikely to make the NativeLibrary callback tests pass on the interpreter, since they were failing at an earlier stage to what I think is a different interpreter bug related to not properly catching exceptions.

File: src/mono/mono/metadata/marshal.c [hunk 1]
```diff
@@ -6755,6 +6752,21 @@ mono_marshal_get_type_object (MonoClass *klass)
...
 
+gpointer
+mono_marshal_lookup_pinvoke (MonoMethod *method)
+{
+	ERROR_DECL (error);
+	gpointer addr;
+
+	g_assert (method);
+	addr = mono_lookup_pinvoke_call_internal (method, error);
+	printf ("%p\n", addr);
```
Comment (lambdageek) [inline src/mono/mono/metadata/marshal.c:6763]: debug printf

**Comment by lambdageek** [comment]
Is there an issue # that this is addressing, or some test that could be re-enabled?

**Comment by vargaz** [comment]
This one:
https://github.com/dotnet/runtime/issues/35219

**Review by CoffeeFlux — APPROVED** [review]
Thanks a ton!

**Comment by SamMonoRT** [comment]
If all these previously failing tests pass please remove from exclusion list in https://github.com/dotnet/runtime/blob/master/src/coreclr/tests/issues.targets

Interop/NativeLibraryResolveCallback/CallbackTests/CallbackTests.sh
Interop/NativeLibraryResolveCallback/CallbackStressTest/CallbackStressTest.sh
Interop/NativeLibraryResolveEvent/ResolveEventTests/ResolveEventTests.sh

**Pull request merged**